src/util/HoldTrack.h

```diff
@@ -0,0 +1,120 @@
+/*
+ * LSST Data Management System
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+#ifndef LSST_QSERV_UTIL_HOLDTRACK_H
+#define LSST_QSERV_UTIL_HOLDTRACK_H
+
+// System headers
+#include <atomic>
+#include <map>
+#include <memory>
+#include <mutex>
+#include <set>
+#include <string>
+#include <thread>
+
+// qserv headers
+#include "util/Issue.h"
+
+// This header declarations
+namespace lsst::qserv::util {
+
+/// This class is used to track where a program is waiting for long periods of time.
+/// It is meant to be used to locate blocks on communication or mutex for very long
+/// periods of time.
+/// For this class to work CheckKeySet() needs to be called on a regular basis, one
+/// call every 5 minutes may be adequate.
+/// Time precision is limited to milliseconds, which could be an issue in loops or
+/// recursion. This can be worked around by changing the note, but that situation
+/// is not where this class is meant to be used.
+class HoldTrack {
+public:
+    using Ptr = std::shared_ptr<HoldTrack>;
+
+    /// Keys sorted by thread::id, sequence_number, uint64_t(time in milliseconds).
+    /// sequence number is needed to keep items in order as too many items may occur in 1ms.
+    using KeyType = std::pair<std::thread::id, std::pair<uint64_t, uint64_t>>;
+
+    HoldTrack() = delete;
+    HoldTrack(HoldTrack const&) = delete;
+    HoldTrack& operator=(HoldTrack const&) = delete;
+
+    ~HoldTrack() = default;
+
+    /// Setup the global instance for storing hold information.
+    /// Items that have existed for longer than `durationLimit` will be logged.
+    static void setup(double durationLimitSeconds);
```

Comment:
This method should use the standard data type:
```
std::chrono::milliseconds
```
If you still like using `double` for some reason within the implementation then this is fine.